[Feature] New JMESPath filter `random()`

### Problem Statement

There have been several requests in the community for a JMESPath filter capable of generating random sequences of characters. This is also a fairly common function found in many other languages/libraries.

### Solution Description

Implement a custom JMESPath `random()` filter that is capable of generating a randomized set of output the length of which, at minimum, can be configurable as a parameter.

### Alternatives

Use another field like UID from a resource to hijack their random string.

### Additional Context

_No response_

### Slack discussion

_No response_

### Research

- [X] I have read and followed the documentation AND the [troubleshooting guide](https://kyverno.io/docs/troubleshooting/).
- [X] I have searched other issues in this repository and mine is not recorded.

@chipzoller Yes I'm working on it. 
I'm getting some errors. In the GetFunctions I have added this. 
```go
		{
			Entry: &gojmespath.FunctionEntry{Name: random,
				Arguments: []ArgSpec{
					{Types: []JpType{JpNumber}},
				},
				Handler: jpRandom,
			},
			ReturnType: []JpType{JpString},
			Note:       "Generates a random sequence of characters",
		},
``` 

Now I'm trying to create a func that will take an integer as argument generate the random string of characters.

```go
func jpRandom(a []interface{}) (interface{}, error) {
	var characters = []rune("abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ")
	b  := make([]rune, a[0].(int))
	for i := range b {
		b[i] = characters[rand.Intn(len(characters))]
	} 
	return string(b),nil
}
```

- When I'm trying to test my function then I'm getting this error. 
```
$ cat random
random(`5`)
$ echo {} | ./kubectl-kyverno jp -e random
panic: interface conversion: interface {} is float64, not int

goroutine 1 [running]:
github.com/kyverno/kyverno/pkg/engine/jmespath.jpRandom({0xc000a0ab60?, 0xc000a0ab60?, 0x1?})
        /home/kranurag7/kyverno/pkg/engine/jmespath/functions.go:899 +0x2dc
github.com/jmespath/go-jmespath.(*functionCaller).CallFunction(0x3631b00?, {0xc000a29280, 0x6}, {0xc000a0ab60, 0x1, 0x1}, 0xc0004aed80)
        /home/kranurag7/go/pkg/mod/github.com/kyverno/go-jmespath@v0.4.1-0.20210511164400-a1d46efa2ed6/functions.go:401 +0x252
github.com/jmespath/go-jmespath.(*treeInterpreter).Execute(0xc0004aed80, {0x4, {0x3544400, 0xc000a0a480}, {0xc00015bdd0, 0x1, 0x1}}, {0x369a600, 0xc00015bd70})
        /home/kranurag7/go/pkg/mod/github.com/kyverno/go-jmespath@v0.4.1-0.20210511164400-a1d46efa2ed6/interpreter.go:93 +0x1fe9
github.com/jmespath/go-jmespath.(*JMESPath).Search(...)
        /home/kranurag7/go/pkg/mod/github.com/kyverno/go-jmespath@v0.4.1-0.20210511164400-a1d46efa2ed6/api.go:37
github.com/kyverno/kyverno/cmd/cli/kubectl-kyverno/jp.Command.func1(0xc0001f7180?, {0xc0006c5f60?, 0x2?, 0x2?})
        /home/kranurag7/kyverno/cmd/cli/kubectl-kyverno/jp/jp_command.go:82 +0x846
github.com/spf13/cobra.(*Command).execute(0xc0001f7180, {0xc0006c5f40, 0x2, 0x2})
        /home/kranurag7/go/pkg/mod/github.com/spf13/cobra@v1.4.0/command.go:856 +0x67cAN
github.com/spf13/cobra.(*Command).ExecuteC(0xc000581b80)
        /home/kranurag7/go/pkg/mod/github.com/spf13/cobra@v1.4.0/command.go:974 +0x3b4
github.com/spf13/cobra.(*Command).Execute(...)
        /home/kranurag7/go/pkg/mod/github.com/spf13/cobra@v1.4.0/command.go:902
main.main()
        /home/kranurag7/kyverno/cmd/cli/kubectl-kyverno/main.go:35 +0xc9
```
I have tried using `n int` as function argument but that's not working out.

--- a/go.mod
+++ b/go.mod
@@ -61,6 +61,7 @@ require (
 
 require (
 	github.com/jmoiron/jsonq v0.0.0-20150511023944-e874b168d07e
+	github.com/zach-klippenstein/goregen v0.0.0-20160303162051-795b5e3961ea
 	go.opentelemetry.io/otel v1.7.0
 	go.opentelemetry.io/otel/exporters/otlp/otlpmetric v0.30.0
 	go.opentelemetry.io/otel/exporters/otlp/otlpmetric/otlpmetricgrpc v0.30.0
@@ -179,6 +180,7 @@ require (
 	github.com/gogo/protobuf v1.3.2 // indirect
 	github.com/golang-jwt/jwt v3.2.2+incompatible // indirect
 	github.com/golang-jwt/jwt/v4 v4.4.2 // indirect
+	github.com/google/gxui v0.0.0-20151028112939-f85e0a97b3a4 // indirect
 	github.com/golang/glog v1.0.0 // indirect
 	github.com/golang/groupcache v0.0.0-20210331224755-41bb18bfe9da // indirect
 	github.com/golang/mock v1.6.0 // indirect

--- a/pkg/engine/jmespath/functions.go
+++ b/pkg/engine/jmespath/functions.go
@@ -5,6 +5,7 @@ import (
 	"encoding/json"
 	"errors"
 	"fmt"
+	"log"
 	"path/filepath"
 	"reflect"
 	"regexp"
@@ -17,6 +18,7 @@ import (
 	"github.com/blang/semver/v4"
 	gojmespath "github.com/jmespath/go-jmespath"
 	wildcard "github.com/kyverno/kyverno/pkg/utils/wildcard"
+	regen "github.com/zach-klippenstein/goregen"
 	"sigs.k8s.io/yaml"
 )
 
@@ -65,6 +67,7 @@ var (
 	parseYAML              = "parse_yaml"
 	items                  = "items"
 	objectFromLists        = "object_from_lists"
+	random                 = "random"
 )
 
 const (
@@ -423,6 +426,16 @@ func GetFunctions() []*FunctionEntry {
 			},
 			ReturnType: []JpType{JpObject},
 			Note:       "converts a pair of lists containing keys and values to an object",
+		},
+		{
+			Entry: &gojmespath.FunctionEntry{Name: random,
+				Arguments: []ArgSpec{
+					{Types: []JpType{JpString}},
+				},
+				Handler: jpRandom,
+			},
+			ReturnType: []JpType{JpString},
+			Note:       "Generates a random sequence of characters",
 		},
 	}
 }
@@ -937,4 +950,17 @@ func validateArg(f string, arguments []interface{}, index int, expectedType refl
 	}
 
 	return arg, nil
+}
+
+func jpRandom(arguments []interface{}) (interface{}, error) {
+	pattern := arguments[0].(string)
+	if pattern == "" {
+		return "", errors.New("no pattern provided")
+	}
+	ans, err := regen.Generate(pattern)
+	if err != nil {
+		fmt.Println("Invalid Pattern: ", err)
+	}
+	log.Println()
+	return ans, nil
 }

--- a/pkg/engine/jmespath/functions.go
+++ b/pkg/engine/jmespath/functions.go
@@ -428,7 +428,8 @@ func GetFunctions() []*FunctionEntry {
 			Note:       "converts a pair of lists containing keys and values to an object",
 		},
 		{
-			Entry: &gojmespath.FunctionEntry{Name: random,
+			Entry: &gojmespath.FunctionEntry{
+				Name: random,
 				Arguments: []ArgSpec{
 					{Types: []JpType{JpString}},
 				},

